apps/prairielearn/src/pages/instructorInstanceAdminPublishing/instructorInstanceAdminPublishing.tsx

```diff
@@ -0,0 +1,244 @@
+      origHash = sha256(
+        b64EncodeUnicode(await fs.readFile(infoCourseInstancePath, 'utf8')),
+      ).toString();
+    }
+
+    const accessRules = await queryRows(
+      sql.course_instance_access_rules,
+      { course_instance_id: res.locals.course_instance.id },
+      CourseInstanceAccessRuleSchema,
+    );
+
+    res.send(
+      PageLayout({
+        resLocals: res.locals,
+        pageTitle: 'Publishing',
+        navContext: {
+          type: 'instructor',
+          page: 'instance_admin',
+          subPage: 'publishing',
+        },
+        content: (
+          <>
+            <CourseInstanceSyncErrorsAndWarnings
+              authzData={res.locals.authz_data}
+              courseInstance={res.locals.course_instance}
+              course={res.locals.course}
+              urlPrefix={res.locals.urlPrefix}
+            />
+            <InstructorInstanceAdminPublishing
+              courseInstance={courseInstance}
+              hasCourseInstancePermissionEdit={hasCourseInstancePermissionEdit}
+              hasCourseInstancePermissionView={hasCourseInstancePermissionView}
+              accessRules={accessRules}
+              csrfToken={res.locals.__csrf_token}
+              origHash={origHash}
+            />
+          </>
+        ),
+      }),
+    );
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    const {
+      authz_data: { has_course_instance_permission_edit: hasCourseInstancePermissionEdit },
+      course_instance: courseInstance,
+    } = extractPageContext(res.locals, {
+      pageType: 'courseInstance',
+      accessType: 'instructor',
+    });
+
+    if (!hasCourseInstancePermissionEdit) {
+      throw new error.HttpStatusError(403, 'Access denied (must be course instance editor)');
+    }
+
+    if (req.body.__action === 'update_publishing') {
+      if (!courseInstance.modern_publishing) {
+        flash('error', 'Cannot update publishing when legacy allowAccess rules are present');
+        res.redirect(req.originalUrl);
+        return;
+      }
+
+      // Read the existing infoCourseInstance.json file
+      const infoCourseInstancePath = path.join(
+        res.locals.course.path,
+        'courseInstances',
+        res.locals.course_instance.short_name,
+        'infoCourseInstance.json',
+      );
+
+      if (!(await fs.pathExists(infoCourseInstancePath))) {
+        flash('error', 'infoCourseInstance.json does not exist');
+        res.redirect(req.originalUrl);
+        return;
+      }
+
+      const courseInstanceInfo: CourseInstanceJsonInput = JSON.parse(
+        await fs.readFile(infoCourseInstancePath, 'utf8'),
+      );
+
+      const parsedResult = z
+        .object({
+          start_date: z.string(),
+          end_date: z.string(),
+        })
+        .safeParse(req.body);
+
+      if (!parsedResult.success) {
+        flash('error', 'Invalid request body');
+        res.redirect(req.originalUrl);
+        return;
+      }
```

Comment:
👀 `safeParse` + `flash(...)` on error is novel, typically we'd just let that 5XX. Say more about introducing this?